### PR TITLE
NetworkInfo - Use a single sortable address type

### PR DESCRIPTION
**Pull request by manadart**
This patch fixes a kluge in `NetworkInfoIAAS` that was used to sort addresses by converting state addresses to `SpaceAddresses` and sorting that as a proxy.

Here, we introduce a new composite type that can be sorted by the now-public order determination function in `core/network`.

Next we will remove the premature materialisation of these addresses into `params.NetworkInfoResult` so that we can sort ingress addresses in addition to bind addresses before populating API returns.

## QA steps

This is a mechanical change verified by passing unit tests.

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[NetworkInfoIAAS - Custom Address Type](https://trello.com/c/nb4VyzOI/3822-networkinfoiaas-custom-address-type)

File: apiserver/facades/agent/uniter/networkinfoiaas.go [hunk 1]
```diff
@@ -18,6 +18,34 @@ import (
...
+type NetInfoAddress interface {
+	network.Address
+
+	// DeviceName is the name of the link-layer device
+	// with which this address is associated.
+	DeviceName() string
+
+	// Device is the link-layer device with which this
+	// address is associated.
+	// TODO (manadart 2021-07-15): Indirect this.
```
Comment (manadart) [inline apiserver/facades/agent/uniter/networkinfoiaas.go:30]: I'll do this in the following patch.

File: apiserver/facades/agent/uniter/networkinfoiaas.go [hunk 2]
```diff
@@ -42,6 +69,23 @@ type machine struct {
...
+		return nil, errors.Trace(err)
+	}
+
+	res := make([]NetInfoAddress, len(addrs))
+	for i, addr := range addrs {
+		spaceAddr, err := network.ConvertToSpaceAddress(addr, subs)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		res[i] = netInfoAddress{addr, spaceAddr}
```
Comment (SimonRichardson) [inline apiserver/facades/agent/uniter/networkinfoiaas.go:84]: nit: can we add the field names here.

File: apiserver/facades/agent/uniter/networkinfoiaas.go [hunk 3]
```diff
@@ -254,41 +298,32 @@ func (n *NetworkInfoIAAS) populateMachineNetworkInfos() error {
...
-		addrByIP[addr.Value()] = addr
+		addrByIP[addr.Host()] = addr
 	}
-
-	spaceAddrs := make(network.SpaceAddresses, len(addrs))
-	for i, addr := range addrs {
-		if spaceAddrs[i], err = network.ConvertToSpaceAddress(addr, n.subs); err != nil {
-			n.populateMachineNetworkInfoErrors(spaceSet, err)
-			return nil
+	sort.Slice(addrs, func(i, j int) bool {
```
Comment (SimonRichardson) [inline apiserver/facades/agent/uniter/networkinfoiaas.go:310]: This is much better.

**Review by SimonRichardson — APPROVED** [review]

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**